### PR TITLE
Reenter after short disconnect

### PR DESCRIPTION
**Pull request by withakay**
possible fix for re-enter presence

### CODE REVIEW

File: src/IO.Ably.Shared/Realtime/Presence.cs [hunk 1]
```diff
@@ -468,49 +467,52 @@ internal void EndSync()
...
+            EnsureLocalPresenceEntered();
+        }
+
+        internal void EnsureLocalPresenceEntered()
+        {
+            foreach (var item in InternalMap.Values)
             {
-                _syncAsResultOfAttach = false;
-                foreach (var item in InternalMap.Values)
+                if (Map.Put(item))
```
[the remark this answers is not in the document]
Comment (withakay) [reply src/IO.Ably.Shared/Realtime/Presence.cs]: So replacing with something like `if (Map.Contains(item))` would be correct?

[the remark this answers is not in the document]
Comment (SimonWoolf) [reply src/IO.Ably.Shared/Realtime/Presence.cs]: Yup. Per https://github.com/ably/docs/pull/711/files: `for each member of the @RTP17@ internal @PresenceMap@ whose @memberKey@ is not also a member of the normal @PresenceMap@, ...`

[the remark this answers is not in the document]
Comment (SimonWoolf) [reply src/IO.Ably.Shared/Realtime/Presence.cs]: (more specifically you're checking if it's _not_ there, so `if(!Map.Contains(item))`)

File: src/IO.Ably.Shared/Realtime/Presence.cs [hunk 2]
```diff
@@ -468,49 +467,52 @@ internal void EndSync()
...
-                    if (Map.Put(item))
+                    var clientId = item.ClientId;
+                    try
                     {
-                        var clientId = item.ClientId;
-                        try
+                        /* Message is new to presence map, send it */
+                        var itemToSend = item.ShallowClone();
+                        itemToSend.ConnectionId = null;
+                        itemToSend.Action = PresenceAction.Enter;
```
[the remark this answers is not in the document]
Comment (withakay) [reply src/IO.Ably.Shared/Realtime/Presence.cs]: yep, I will replace with 
`var itemToSend = new PresenceMessage(PresenceAction.Enter, item.ClientId, item.Data);`

File: src/IO.Ably.Shared/Realtime/Presence.cs [hunk 3]
```diff
@@ -468,49 +467,52 @@ internal void EndSync()
...
+                            $"Cannot automatically re-enter {clientId} on channel {_channel.Name} ({e.ErrorInfo.Message})";
+                        Logger.Error(errorString);
+                        _channel.EmitUpdate(new ErrorInfo(errorString, 91004), true);
                     }
                 }
-
-                InternalMap.Clear();
             }
+
+            InternalMap.Clear();
```
[the remark this answers is not in the document]
Comment (withakay) [reply src/IO.Ably.Shared/Realtime/Presence.cs]: fixed by https://github.com/ably/ably-dotnet/pull/358/commits/1395f0a4830582f6d3eb80effdbd65af0a861b71

File: src/IO.Ably.Shared/Realtime/Presence.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (withakay) [reply src/IO.Ably.Shared/Realtime/Presence.cs]: fixed by https://github.com/ably/ably-dotnet/pull/358/commits/1395f0a4830582f6d3eb80effdbd65af0a861b71

File: src/IO.Ably.Shared/Realtime/Presence.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (withakay) [reply src/IO.Ably.Shared/Realtime/Presence.cs]: fixed by https://github.com/ably/ably-dotnet/pull/358/commits/1395f0a4830582f6d3eb80effdbd65af0a861b71

**Comment by withakay** [comment]
This branch needs some of tests adding covering RTP17e and tests to simulate the manual testing of pausing and resuming windows for ~1 minute and also for > 2 minutes. @marto83 is going to have a look at those.

**Pull request merged**